core/src/Revolution/Services/NotFoundException.php

```diff
@@ -0,0 +1,13 @@
+<?php
+
+
+namespace MODX\Revolution\Services;
+
+
+use Exception;
+use Psr\Container\NotFoundExceptionInterface;
+
+class NotFoundException extends Exception implements NotFoundExceptionInterface
```

Comment:
Perhaps this should extend the `ContainerException`, so a single try/catch is enough to wrap a call to `Container->get()`

Reply:
https://www.php-fig.org/psr/psr-11/#33-psrcontainernotfoundexceptioninterface

Reply:
See: https://www.php-fig.org/psr/psr-11/#31-psrcontainercontainerinterface

Reply:
I meant for the MODX-specific implementation (class), though it's good to know the interface does extend.

Reply:
Both exceptions have a different purpose:
```
 * @throws NotFoundExceptionInterface  No entry was found for **this** identifier.
 * @throws ContainerExceptionInterface Error while retrieving the entry.
```

Reply:
So can you catch `ContainerExceptionInterface`? Or catch `\Exception` and test if it is an instance of `ContainerExceptionInterface`?

Reply:
I'm mostly thinking along these lines:

```
try {
   $someservice = $container->get('someservice');
} catch (??? $e) {
   log('install someservice first or whatever');
}
```

In such a case, I don't care wether it's a not found or an internal error to the container or an error in the instantiation of the dependency - all I want to know is, did the call succeed, was my service available.

Using `ContainerExceptionInterface` would definitely work, but that makes the code reliant on the dependency. If we're making available MODX-specific implementations, it would seem safer to look for that specifically. But in the current situation that would mean either doubling up the `catch` to cover both cases (which causes the boilerplate code to double in size for each service you're looking up), or catching the root `\Exception` (which seems like a waste of more detailed implementations). 

I don't necessarily have the answer but think this is the type of architectural detail that's worth thinking about, as it could become a pattern that is implemented into other code as it gets refactored. 

My gut says NotFoundException should be an instance of ContainerException so they can be caught in a single catch statement, and that we should use that instead of code from a dependency directly. I've had issues in Commerce where a dependency update caused breaking changes to extensions because it was exposed directly to users; wrapping that in a Commerce-specific class helps avoid such lock in in the future.

But, as the dependency here is a well-established PSR standard, I can also imagine we'd prefer referring to those directly.

Reply:
Yes, I want to stick to the PSR standard. We are 100% in control of the PSR-11 extension of Pimple here. I can guarantee that the PSR-11 interfaces here will never change. They will use proper semver if they do and it will not affect us unless we choose to adopt them.

As for NotFoundException extending ContainerException, can I even do that if they both implement ContainerInterfaceException? I'm not even sure if that is proper architecture.

Reply:
If you don't care about it being a `NotFoundExceptionInterface` or `ContainerExceptionInterface` why not simply catch `Exception`?

Reply:
Any consensus here? Can we move forward with this as-is and address the concern with the exceptions post-haste?

Reply:
I'd go with as-is. Catching `\Exception` seems fine to me.

Reply:
I'm okay with catching the PSR interfaces, but want to play with the PR a little before giving formal approval if that's alright.

Reply:
I'll give you two hours! That enough? LOL!

Reply:
Ha-ha. If we're going to use this as a primary integration method for the next 5-10 years, why rush?

Reply:
One of my PRs is also delayed because we need to come to a consensus first 😅😋

Reply:
Because no one can actually build an extra for 3.x without this?

Reply:
Well thumbs down to you to, buddy! 😅